### PR TITLE
feat: sdk 712 work manager support

### PR DESCRIPTION
**Pull request by itsdebs**
**Fixes** # (*issue*)

## What
Moved Work Manager to be a separate library.
Configuration `isPeriodicFlushEnabled` needs to be true to turn on Work manager.
Default time is 15 minutes. This is not configurable. (We can make this configurable based on discussion)

**How to implement**

- Users need to implement sync library(Gradle details to be provided later)
- Extend `WorkManagerAnalyticsFactory` class. This class instance will be created through reflection, hence it should have a default instance. This instance will provide an `Analytics` object. This is done to ensure that in case the work runs when Analytics instance has been garbage collected, we can create a new one through this factory.

```
class SampleWorkManagerAnalyticsFactory : WorkManagerAnalyticsFactory {
    override fun createAnalytics(application: Application): Analytics {
        return RudderAnalyticsUtils.createAnalyticsInstance(application)
    }
}
```
- Extend `WorkManagerPlugin` class to provide the class name for the factory
```
class SampleWorkManagerPlugin: WorkerManagerPlugin() {
    override val workManagerAnalyticsFactoryClassName: Class<out WorkManagerAnalyticsFactory>
        get() = SampleWorkManagerAnalyticsFactory::class.java
}
```
- Create the Analytics instance
```
    fun createAnalyticsInstance(application: Application): Analytics {
        return RudderAnalytics(
            writeKey = WRITE_KEY,
            initializationListener = { success, message ->
                // DIY
            },
            configuration = ConfigurationAndroid(
                application = application,
                GsonAdapter(),
                dataPlaneUrl = DATA_PLANE_URL,
                controlPlaneUrl = CONTROL_PLANE_URL,
                trackLifecycleEvents = true,
                recordScreenViews = true,
                isPeriodicFlushEnabled = true

            )
        )
    }
```    
- Add the extended WorkManagerPlugin to Analytics as `InfrastructurePlugin`
 ```
analytics.addInfrastructurePlugin(SampleWorkManagerPlugin())
```

## Why
- A separate library will reduce complexity in main code. 
- Work manager features/bug fixes can be solved independently
- Reduce code weight. 
- The factory to initialize Analytics is required since we don't want to store instance of Analytics from application as it might lead to memory leaks.
- Reflection is necessary to initialize the factory as we can only pass in simple data to the Worker class. 

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
## Checklist:
- [ ] Version upgraded (project, README, gradle, podspec etc)
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by ChryssaAliferi — APPROVED** [review]
Some minor comments, it would be great to check them out.
Nice work!!

File: samples/sample-kotlin-android/src/main/kotlin/com/rudderstack/android/sampleapp/mainview/MainViewModel.kt [hunk 1]
```diff
@@ -30,17 +30,17 @@ class MainViewModel(application: Application) : AndroidViewModel(application) {
 
     private val _loggingInterceptor by lazy {
         object : Plugin {
-            private var _instanceName: String? = null
+            private var _writeKey: String? = null
```
[the remark this answers is not in the document]
Comment (itsdebs) [reply samples/sample-kotlin-android/src/main/kotlin/com/rudderstack/android/sampleapp/mainview/MainViewModel.kt:33]: I think this is already changed in feat/android branch

File: samples/sample-kotlin-android/src/main/kotlin/com/rudderstack/android/sampleapp/analytics/RudderAnalyticsUtils.kt [hunk 2]
```diff
@@ -62,6 +47,29 @@ object RudderAnalyticsUtils {
                 )
             ), JacksonAdapter()
         )
+        _rudderAnalytics?.initializeWorkManager()
+    }
+
+    fun createV1AnalyticsInstance(application: Application, listener: InitializationListener? = null): Analytics {
```
[the remark this answers is not in the document]
Comment (itsdebs) [reply samples/sample-kotlin-android/src/main/kotlin/com/rudderstack/android/sampleapp/analytics/RudderAnalyticsUtils.kt:53]: Renaming for better meaning.

File: libs/sync/src/main/java/com/rudderstack/android/sync/internal/RudderSyncWorker.kt [hunk 3]
```diff
@@ -12,37 +12,45 @@
...
+                                                    "already shutdown")
+                return Result.failure()
+            }
+            val sinkAnalytics = (weakSinkAnalytics ?: createSinkAnalytics())
             val success = sinkAnalytics?.blockingFlush()
             sinkAnalytics?.logger?.debug(log = "Data upload through worker. success: $success")
-            if(weakSinkAnalytics == null)
-                sinkAnalytics?.shutdown()
-            return if(success == true) Result.success() else Result.failure()
+            if (weakSinkAnalytics == null) sinkAnalytics?.shutdown()
```
[the remark this answers is not in the document]
Comment (itsdebs) [reply libs/sync/src/main/java/com/rudderstack/android/sync/internal/RudderSyncWorker.kt:51]: Sometimes if-else is used as replacement for missing ternary operators, do you think we should use braces for those too?

File: libs/sync/src/test/java/com/rudderstack/android/sync/RudderSyncWorkerTest.kt
Comment (1abhishekpandey) [inline libs/sync/src/test/java/com/rudderstack/android/sync/RudderSyncWorkerTest.kt:1]: There is an unused import directive, please remove it in future PR:
> import com.rudderstack.android.sync.internal.registerWorkManager

File: libs/sync/src/test/java/com/rudderstack/android/sync/RudderSyncWorkerTest.kt [hunk 4]
```diff
@@ -68,22 +69,31 @@ class RudderSyncWorkerTest {
         every{analytics.logger} returns logger
         every{analytics.currentConfigurationAndroid} returns configuration
         every{analytics.currentConfiguration} returns configuration
-        application.registerWorkManager(analytics, DummyAnalyticsFactory::class.java)
+        every { analytics.writeKey } returns "test_write_key"
+        every { analytics.shutdown() } just runs
+//        application.registerWorkManager(analytics, DummyAnalyticsFactory::class.java)
```
Comment (1abhishekpandey) [inline libs/sync/src/test/java/com/rudderstack/android/sync/RudderSyncWorkerTest.kt:74]: Please remove this comment.

File: libs/sync/src/main/java/com/rudderstack/android/sync/WorkerManagerPlugin.kt [hunk 5]
```diff
@@ -35,9 +39,16 @@ abstract class WorkerManagerPlugin : InfrastructurePlugin {
...
 
+    /**
+     * Internal classes are not supported.
+     * This is because instantiating an inner class requires the parent class instance.
+     * It's not worth it to try finding an instance in Heap. Cause this approach might conflict
+     * with garbage collector
+     */
     abstract val workManagerAnalyticsFactoryClassName: Class<out WorkManagerAnalyticsFactory>
 
     override fun updateConfiguration(configuration: Configuration) {
```
[the remark this answers is not in the document]
Comment (itsdebs) [reply libs/sync/src/main/java/com/rudderstack/android/sync/WorkerManagerPlugin.kt:54]: No, then user has to implement it positively

[the remark this answers is not in the document]
Comment (1abhishekpandey) [reply libs/sync/src/main/java/com/rudderstack/android/sync/WorkerManagerPlugin.kt:54]: Just to understand this `updateConfiguration()` is an optional method, right?
I attempted to comment out this code and then executed it. It didn't prompt me to implement this method. Could you please verify if this is the case or not?

I think we can safely remove this `updateConfiguration()`, as it is an optional method. You may verify it from the corresponding interface class. Please correct me, if I'm mistaken.

File: samples/sample-kotlin-android/build.gradle.kts [hunk 6]
```diff
@@ -149,8 +150,10 @@ dependencies {
     implementation(project(":core"))
     implementation(project(":models"))
     implementation(project(":web"))
-    implementation(project(":rudderjsonadapter"))
+//    implementation(project(":rudderjsonadapter"))
```
Comment (1abhishekpandey) [inline samples/sample-kotlin-android/build.gradle.kts:153]: Please remove this if not needed.

File: samples/sample-kotlin-android/build.gradle.kts [hunk 7]
```diff
@@ -110,11 +111,11 @@ android {
     buildFeatures {
         buildConfig = true
     }
-    packagingOptions {
+/*    packagingOptions {
         resources {
             excludes += "/META-INF/{AL2.0,LGPL2.1}"
         }
-    }
+    }*/
```
Comment (1abhishekpandey) [inline samples/sample-kotlin-android/build.gradle.kts:118]: Please consider removing this comment block as well, if not needed.

File: libs/navigationplugin/build.gradle.kts [hunk 8]
```diff
@@ -63,3 +63,6 @@ dependencies {
 tasks.withType(type = org.jetbrains.kotlin.gradle.tasks.KaptGenerateStubs::class) {
     kotlinOptions.jvmTarget = "17"
 }
+apply(from = "${project.projectDir.parentFile.parent}/gradle/artifacts-aar.gradle")
+apply(from = "${project.projectDir.parentFile.parent}/gradle/mvn-publish.gradle")
+apply(from = "${project.projectDir.parentFile.parent}/gradle/codecov.gradle")
```
Comment (ChryssaAliferi) [inline libs/navigationplugin/build.gradle.kts:68]: 👍 Amazing!!

**Review by ChryssaAliferi — APPROVED** [review]
👍 Nice work!!

File: libs/navigationplugin/build.gradle.kts (hunk 8 above)
Comment (ChryssaAliferi) [reply libs/navigationplugin/build.gradle.kts:68]: This is a refactor we need to do on other modules as well 👏

**Pull request merged**